File: goalsetter-rag-starter-frontend/src/services/goalService.ts
// src/services/goalService.ts

import api from '../api';
import { Goal } from '../models/Goal';
import { Tag } from '../models/Tag';
import { StrapiServiceResponse } from '../types/StrapiServiceResponse';
import { ErrorResponse } from '../types/ErrorResponse';

const goalService = {
  getAll: async (keycloakSubject: string | undefined): Promise<StrapiServiceResponse<Goal[]>> => {
    try {
      if (!keycloakSubject) {
        const result: StrapiServiceResponse<Goal[]> = {
          data: null,
          error: {
            status: 400,
            statusText: 'Bad Request',
            data: { message: 'Goal ID is undefined' },
          },
        };
        return result;
      }
      const response = await api.get(`/goals?filters[userProfile][keycloaksubject][$eq]=${keycloakSubject}`);
      const result: StrapiServiceResponse<Goal[]> = { data: response.data.data, error: null };
      return result;
    } catch (error) {
      const result: StrapiServiceResponse<Goal[]> = { data: null, error: error as ErrorResponse };
      return result;
    }
  },
  get: async (id: number | undefined): Promise<StrapiServiceResponse<Goal>> => {
    try {

      if (!id) {
        const result: StrapiServiceResponse<Goal> = {
          data: null,
          error: {
            status: 400,
            statusText: 'Bad Request',
            data: { message: 'Goal ID is undefined' },
          },
        };
        return result;
      }

      const response = await api.get(`/goals/${id}?populate[0]=tags&populate[1]=userProfile`);
      const result: StrapiServiceResponse<Goal> = { data: response.data.data, error: null };
      return result;
    } catch (error) {
      const result: StrapiServiceResponse<Goal> = { data: null, error: error as ErrorResponse };
      return result;
    }
  },
  getByTags: async (tags: Tag[]): Promise<StrapiServiceResponse<Goal[]>> => {
    try {
      if (!tags) {
        const result: StrapiServiceResponse<Goal[]> = {
          data: null,
          error: {
            status: 400,
            statusText: 'Bad Request',
            data: { message: 'tags undefined' },
          },
        };
        return result;
      }
  
      const tagFilters = tags
        .map((tag) => `filters[tags][id][$in]=${tag.id}`)
        .join('&');
  
      const response = await api.get(`/goals?${tagFilters}`);
      const result: StrapiServiceResponse<Goal[]> = { data: response.data.data, error: null };
      return result;
    } catch (error) {
      const result: StrapiServiceResponse<Goal[]> = { data: null, error: error as ErrorResponse };
      return result;
    }
  },  
  create: async (goal: Goal): Promise<StrapiServiceResponse<Goal>> => {
    try {
      const response = await api.post('/goals', { data: goal });
      const result: StrapiServiceResponse<Goal> = { data: response.data.data, error: null };
      return result;
    } catch (error) {
      const result: StrapiServiceResponse<Goal> = { data: null, error: error as ErrorResponse };
      return result;
    }
  },
  update: async (id: string | undefined, goal: Goal): Promise<StrapiServiceResponse<Goal>> => {
    try {

      if (!id) {
        const result: StrapiServiceResponse<Goal> = {
          data: null,
          error: {
            status: 400,
            statusText: 'Bad Request',
            data: { message: 'Goal ID is undefined' },
          },
        };
        return result;
      }

      const response = await api.put(`/goals/${id}`, { data: goal });
      const result: StrapiServiceResponse<Goal> = { data: response.data.data, error: null };
      return result;
    } catch (error) {
      const result: StrapiServiceResponse<Goal> = { data: null, error: error as ErrorResponse };
      return result;
    }
  },
  delete: async (id: number | undefined): Promise<StrapiServiceResponse<boolean>> => {
    try {

      if (!id) {
        const result: StrapiServiceResponse<boolean> = {
          data: null,
          error: {
            status: 400,
            statusText: 'Bad Request',
            data: { message: 'Goal ID is undefined' },
          },
        };
        return result;
      }

      await api.delete(`/goals/${id}`);
      const result: StrapiServiceResponse<boolean> = { data: true, error: null };
      return result;
    } catch (error) {
      const result: StrapiServiceResponse<boolean> = { data: false, error: error as ErrorResponse };
      return result;
    }
  },
};

export default goalService;

File: quarkus-embeddings/src/main/java/com/enterprisewebservice/api/GoalService.java
package com.enterprisewebservice.api;

import com.enterprisewebservice.model.ErrorResponse;
import com.enterprisewebservice.model.Goal;
import com.enterprisewebservice.model.StrapiServiceResponse;
import com.enterprisewebservice.model.Tag;
import com.fasterxml.jackson.databind.ObjectMapper;

import jakarta.enterprise.context.ApplicationScoped;
import jakarta.inject.Inject;
import java.io.IOException;
import java.util.List;

@ApplicationScoped
public class GoalService {

    @Inject
    ApiClient apiClient;
    
    private final ObjectMapper objectMapper = new ObjectMapper(); // replace with your actual JSON library

    public StrapiServiceResponse<List<Goal>> getAll(String keycloakSubject) throws IOException {
        if (keycloakSubject == null || keycloakSubject.isEmpty()) {
            return new StrapiServiceResponse<>(null, new ErrorResponse(400, "Bad Request", "Goal ID is undefined"));
        }
        String response = null;
        try {
            response = apiClient.get("/goals?filters[userProfile][keycloaksubject][$eq]=" + keycloakSubject);
        } catch (Exception e) {
            // TODO Auto-generated catch block
            e.printStackTrace();
        }
        return objectMapper.readValue(response, objectMapper.getTypeFactory().constructParametricType(StrapiServiceResponse.class, objectMapper.getTypeFactory().constructCollectionType(List.class, Goal.class)));

    }

    public StrapiServiceResponse<Goal> getGoal(Integer id) throws IOException {
        if (id == null) {
            return new StrapiServiceResponse<>(null, new ErrorResponse(400, "Bad Request", "Goal ID is undefined"));
        }
        String response = null;
        try {
            response = apiClient.get("/goals?populate[0]=tags&populate[1]=userProfile");
        } catch (Exception e) {
            // TODO Auto-generated catch block
            e.printStackTrace();
        }
        return objectMapper.readValue(response, StrapiServiceResponse.class);
    }

    public StrapiServiceResponse<List<Goal>> getByTags(List<Tag> tags, String keycloakSubject) throws IOException {
        if (tags == null || tags.isEmpty()) {
            return new StrapiServiceResponse<>(null, new ErrorResponse(400, "Bad Request", "Tags are undefined"));
        }
        StringBuilder tagFilters = new StringBuilder();
        for (Tag tag : tags) {
            tagFilters.append("filters[tags][id][$in]=").append(tag.getId()).append("&");
        }
        tagFilters.deleteCharAt(tagFilters.length() - 1); // Remove the last "&"
        String response = null;
        try {
            response = apiClient.get("/goals?filters[userProfile][keycloaksubject][$eq]=" + keycloakSubject + "&" + tagFilters.toString());
        } catch (Exception e) {
            // TODO Auto-generated catch block
            e.printStackTrace();
        }
        return objectMapper.readValue(response, objectMapper.getTypeFactory().constructParametricType(StrapiServiceResponse.class, objectMapper.getTypeFactory().constructCollectionType(List.class, Goal.class)));

    }

    // Note: For methods requiring POST, PUT and DELETE, you will have to implement similar methods in ApiClient and use them here.
}
